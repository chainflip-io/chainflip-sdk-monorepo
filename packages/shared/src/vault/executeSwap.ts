import { ContractReceipt } from 'ethers';
import { Vault__factory } from '../abis';
import {
  checkAllowance,
  getTokenContractAddress,
  getVaultManagerContractAddress,
} from '../contracts';
import { assetContractIds, chainContractIds } from '../enums';
import { assert, isTokenCall, isTokenSwap } from '../guards';
import {
  executeOptionsSchema,
  type ExecuteOptions,
  type ExecuteSwapParams,
  executeSwapParamsSchema,
  type NativeSwapParams,
  type TokenSwapParams,
  NativeCallParams,
  TokenCallParams,
} from './schemas';

const swapNative = async (
  { destChain, destAsset, destAddress, amount }: NativeSwapParams,
  { network, vaultContractAddress: address, signer, ...opts }: ExecuteOptions,
): Promise<ContractReceipt> => {
  const vaultContractAddress =
    network === 'localnet' ? address : getVaultManagerContractAddress(network);

  const vault = Vault__factory.connect(vaultContractAddress, signer);

  const transaction = await vault.xSwapNative(
    chainContractIds[destChain],
    destAddress,
    assetContractIds[destAsset],
    [],
    { value: amount, ...opts },
  );

  return transaction.wait(1);
};

const swapToken = async (
  params: TokenSwapParams,
  {
    network,
    vaultContractAddress: vaultAddress,
    srcTokenContractAddress: tokenAddress,
    signer,
    ...opts
  }: ExecuteOptions,
): Promise<ContractReceipt> => {
  const vaultContractAddress =
    network === 'localnet'
      ? vaultAddress
      : getVaultManagerContractAddress(network);

  const erc20Address =
    network === 'localnet'
      ? tokenAddress
      : getTokenContractAddress(params.srcAsset, network);

  assert(erc20Address !== undefined, 'Missing ERC20 contract address');

  const { isAllowable } = await checkAllowance(
    params.amount,
    vaultContractAddress,
    erc20Address,
    signer,
  );
  assert(isAllowable, 'Swap amount exceeds allowance');

  const vault = Vault__factory.connect(vaultContractAddress, signer);

  const transaction = await vault.xSwapToken(
    chainContractIds[params.destChain],
    params.destAddress,
    assetContractIds[params.destAsset],
    erc20Address,
    params.amount,
    [],
    opts,
  );

  return transaction.wait(1);
};

const callNative = async (
  params: NativeCallParams,
  {
    network,
    vaultContractAddress: vaultAddress,
    signer,
    ...opts
  }: ExecuteOptions,
): Promise<ContractReceipt> => {
  const vaultContractAddress =
    network === 'localnet'
      ? vaultAddress
      : getVaultManagerContractAddress(network);

  const vault = Vault__factory.connect(vaultContractAddress, signer);

  const transaction = await vault.xCallNative(
    chainContractIds[params.destChain],
    params.destAddress,
    assetContractIds[params.destAsset],
    params.ccmMetadata.message,
    params.ccmMetadata.gasBudget,
    [],
    { value: params.amount, ...opts },
  );

  return transaction.wait(1);
};

const callToken = async (
  params: TokenCallParams,
  {
    signer,
    network,
    vaultContractAddress: vaultAddress,
    srcTokenContractAddress: tokenAddress,
    ...opts
  }: ExecuteOptions,
): Promise<ContractReceipt> => {
  const vaultContractAddress =
    network === 'localnet'
      ? vaultAddress
      : getVaultManagerContractAddress(network);

  const erc20Address =
    network === 'localnet'
      ? tokenAddress
      : getTokenContractAddress(params.srcAsset, network);

  assert(erc20Address !== undefined, 'Missing ERC20 contract address');

  const { isAllowable } = await checkAllowance(
    params.amount,
    vaultContractAddress,
    erc20Address,
    signer,
  );
  assert(isAllowable, 'Swap amount exceeds allowance');

  const vault = Vault__factory.connect(vaultContractAddress, signer);

  const transaction = await vault.xCallToken(
    chainContractIds[params.destChain],
    params.destAddress,
    assetContractIds[params.destAsset],
    params.ccmMetadata.message,
    params.ccmMetadata.gasBudget,
    erc20Address,
    params.amount,
    [],
    opts,
  );

  return transaction.wait(1);
};

const executeSwap = async (
  params: ExecuteSwapParams,
  options: ExecuteOptions,
): Promise<ContractReceipt> => {
  const parsedParams = executeSwapParamsSchema.parse(params);
  const opts = executeOptionsSchema.parse(options);

  if ('ccmMetadata' in parsedParams) {
    return isTokenCall(parsedParams)
      ? callToken(parsedParams, opts)
      : callNative(parsedParams, opts);
  }

  return isTokenSwap(parsedParams)
    ? swapToken(parsedParams, opts)
    : swapNative(parsedParams, opts);
};

export default executeSwap;
