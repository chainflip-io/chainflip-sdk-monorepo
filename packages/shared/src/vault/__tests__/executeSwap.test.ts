/* eslint-disable @typescript-eslint/no-empty-function */
/* eslint-disable @typescript-eslint/lines-between-class-members */
/* eslint-disable max-classes-per-file */
import { VoidSigner } from 'ethers';
import { Assets, ChainflipNetworks, Chains } from '../../enums';
import executeSwap from '../executeSwap';
import { ExecuteSwapParams } from '../index';

const ETH_ADDRESS = '0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2';
const DOT_ADDRESS = '5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX';
const REGTEST_BTC_ADDRESS =
  'bcrt1p7mnll6qup4l3lcggvh7t26m4gryawjy0d0cganzh6e4zjm2d3rtqt9usqx';
const TESTNET_BTC_ADDRESS = 'tb1qge9vvd2mmjxfhuxuq204h4fxxphr0vfnsnx205';
const MAINNET_BTC_ADDRESS =
  'bc1pv7lmxr8vvf220cumd4pft77l4pds85pt4l6rw6yrr3cghyf5kl7sq76puk';

class MockVault {
  constructor(readonly address: string) {}
  async xSwapNative(): Promise<any> {}
  async xSwapToken(): Promise<any> {}
  async xCallNative(): Promise<any> {}
  async xCallToken(): Promise<any> {}
}

class MockERC20 {
  async approve(): Promise<any> {}
  async allowance(): Promise<any> {
    return BigInt(Number.MAX_SAFE_INTEGER - 1);
  }
}

jest.mock('../../abis/factories/Vault__factory', () => ({
  Vault__factory: class {
    static connect(address: string) {
      return new MockVault(address);
    }
  },
}));

jest.mock('../../abis/factories/ERC20__factory', () => ({
  ERC20__factory: class {
    static connect() {
      return new MockERC20();
    }
  },
}));

describe(executeSwap, () => {
  it.each([
    {
      destAsset: Assets.BTC,
      destChain: Chains.Bitcoin,
      destAddress: TESTNET_BTC_ADDRESS,
      srcAsset: Assets.ETH,
      srcChain: Chains.Ethereum,
    },
    {
      destAsset: 'BTC',
      destChain: 'Bitcoin',
      destAddress: TESTNET_BTC_ADDRESS,
      srcAsset: Assets.ETH,
      srcChain: Chains.Ethereum,
    },
    {
      destAsset: Assets.FLIP,
      destChain: Chains.Ethereum,
      destAddress: ETH_ADDRESS,
      srcAsset: Assets.ETH,
      srcChain: Chains.Ethereum,
    },
    {
      destAsset: Assets.USDC,
      destChain: Chains.Ethereum,
      destAddress: ETH_ADDRESS,
      srcAsset: Assets.ETH,
      srcChain: Chains.Ethereum,
    },
    {
      destAsset: Assets.DOT,
      destChain: Chains.Polkadot,
      destAddress: DOT_ADDRESS,
      srcAsset: Assets.ETH,
      srcChain: Chains.Ethereum,
    },
  ] as Omit<ExecuteSwapParams, 'amount'>[])(
    'submits a native swap (%p)',
    async (params) => {
      const wait = jest
        .fn()
        .mockResolvedValue({ status: 1, hash: 'hello world' });
      const swapSpy = jest
        .spyOn(MockVault.prototype, 'xSwapNative')
        .mockResolvedValue({ wait });

      expect(
        await executeSwap(
          { amount: '1', ...params } as ExecuteSwapParams,
          {
            network: ChainflipNetworks.sisyphos,
            signer: new VoidSigner('MY ADDRESS').connect({
              getNetwork: () => Promise.resolve({ chainId: 11155111n }),
            } as any),
          },
          {},
        ),
      ).toStrictEqual({ status: 1, hash: 'hello world' });
      expect(wait).toHaveBeenCalledWith(undefined);
      expect(swapSpy.mock.calls).toMatchSnapshot();
    },
  );

  it.each([
    ...[
      { srcAsset: Assets.FLIP, srcChain: Chains.Ethereum },
      { srcAsset: Assets.USDC, srcChain: Chains.Ethereum },
    ].flatMap((src) => [
      {
        destAsset: Assets.BTC,
        destChain: Chains.Bitcoin,
        destAddress: MAINNET_BTC_ADDRESS,
        ...src,
      },
      {
        destAsset: 'BTC',
        destChain: 'Bitcoin',
        destAddress: MAINNET_BTC_ADDRESS,
        ...src,
      },
      {
        destAsset: Assets.ETH,
        destChain: Chains.Ethereum,
        destAddress: ETH_ADDRESS,
        ...src,
      },
      {
        destAsset: Assets.DOT,
        destChain: Chains.Polkadot,
        destAddress: DOT_ADDRESS,
        ...src,
      },
    ]),
  ] as Omit<ExecuteSwapParams, 'amount'>[])(
    'submits a token swap (%p)',
    async (params) => {
      const wait = jest
        .fn()
        .mockResolvedValue({ status: 1, hash: 'hello world' });
      const approveSpy = jest
        .spyOn(MockERC20.prototype, 'approve')
        .mockResolvedValue({ wait });
      const swapSpy = jest
        .spyOn(MockVault.prototype, 'xSwapToken')
        .mockResolvedValue({ wait });
      const allowanceSpy = jest
        .spyOn(MockERC20.prototype, 'allowance')
        .mockResolvedValueOnce(BigInt(Number.MAX_SAFE_INTEGER - 1));

      expect(
        await executeSwap(
          { amount: '1', ...params } as ExecuteSwapParams,
          {
            network: 'mainnet',
            signer: new VoidSigner('MY ADDRESS').connect({
              getNetwork: () => Promise.resolve({ chainId: 1n }),
            } as any),
          },
          {},
        ),
      ).toStrictEqual({ status: 1, hash: 'hello world' });
      expect(wait).toHaveBeenCalledWith(undefined);
      expect(swapSpy.mock.calls).toMatchSnapshot();
      expect(allowanceSpy.mock.calls).toMatchSnapshot();
      expect(approveSpy).not.toHaveBeenCalled();
    },
  );

  it('submits a token swap with sufficient approval', async () => {
    const wait = jest
      .fn()
      .mockResolvedValue({ status: 1, hash: 'hello world' });
    const approveSpy = jest
      .spyOn(MockERC20.prototype, 'approve')
      .mockRejectedValue(Error('unmocked call'));
    const swapSpy = jest
      .spyOn(MockVault.prototype, 'xSwapToken')
      .mockResolvedValue({ wait });
    const allowanceSpy = jest
      .spyOn(MockERC20.prototype, 'allowance')
      .mockResolvedValueOnce(BigInt(Number.MAX_SAFE_INTEGER - 1));

    expect(
      await executeSwap(
        {
          destAsset: Assets.BTC,
          destChain: Chains.Bitcoin,
          destAddress: TESTNET_BTC_ADDRESS,
          srcAsset: Assets.FLIP,
          srcChain: Chains.Ethereum,
          amount: '1',
        },
        {
          network: 'sisyphos',
          signer: new VoidSigner('MY ADDRESS').connect({
            getNetwork: () => Promise.resolve({ chainId: 11155111n }),
          } as any),
        },
        {},
      ),
    ).toStrictEqual({ status: 1, hash: 'hello world' });
    expect(wait).toHaveBeenCalledWith(undefined);
    expect(swapSpy.mock.calls).toMatchSnapshot();
    expect(allowanceSpy.mock.calls).toMatchSnapshot();
    expect(approveSpy).not.toHaveBeenCalled();
  });

  it('can be invoked with localnet options', async () => {
    const wait = jest
      .fn()
      .mockResolvedValue({ status: 1, hash: 'hello world' });
    const approveSpy = jest
      .spyOn(MockERC20.prototype, 'approve')
      .mockResolvedValue({ wait });
    const swapSpy = jest
      .spyOn(MockVault.prototype, 'xSwapToken')
      .mockResolvedValue({ wait });
    const allowanceSpy = jest
      .spyOn(MockERC20.prototype, 'allowance')
      .mockResolvedValueOnce(BigInt(Number.MAX_SAFE_INTEGER - 1));

    expect(
      await executeSwap(
        {
          destAsset: Assets.BTC,
          destChain: Chains.Bitcoin,
          destAddress: REGTEST_BTC_ADDRESS,
          srcAsset: Assets.FLIP,
          amount: '1',
          srcChain: Chains.Ethereum,
        },
        {
          network: 'localnet',
          signer: new VoidSigner('MY ADDRESS'),
          vaultContractAddress: '0x123',
          srcTokenContractAddress: '0x456',
        },
        {},
      ),
    ).toStrictEqual({ status: 1, hash: 'hello world' });
    expect(wait).toHaveBeenCalledWith(undefined);
    expect(swapSpy.mock.calls).toMatchSnapshot();
    expect(allowanceSpy.mock.calls).toMatchSnapshot();
    expect(approveSpy).not.toHaveBeenCalled();
  });

  it('accepts a nonce', async () => {
    const wait = jest
      .fn()
      .mockResolvedValue({ status: 1, hash: 'hello world' });
    const swapSpy = jest
      .spyOn(MockVault.prototype, 'xSwapNative')
      .mockResolvedValue({ wait });

    expect(
      await executeSwap(
        {
          amount: '1',
          destAsset: Assets.BTC,
          destChain: Chains.Bitcoin,
          destAddress: TESTNET_BTC_ADDRESS,
          srcAsset: Assets.ETH,
          srcChain: Chains.Ethereum,
        },
        {
          network: ChainflipNetworks.sisyphos,
          signer: new VoidSigner('MY ADDRESS').connect({
            getNetwork: () => Promise.resolve({ chainId: 11155111n }),
          } as any),
        },
        { nonce: 1 },
      ),
    ).toStrictEqual({ status: 1, hash: 'hello world' });
    expect(wait).toHaveBeenCalledWith(undefined);
    expect(swapSpy.mock.calls).toMatchSnapshot();
  });

  it('rejects if source asset and chain are not valid', async () => {
    await expect(
      executeSwap(
        {
          amount: '1',
          destAsset: Assets.BTC,
          destChain: Chains.Bitcoin,
          destAddress: TESTNET_BTC_ADDRESS,
          srcAsset: Assets.BTC,
          srcChain: Chains.Ethereum,
        },
        {
          network: ChainflipNetworks.sisyphos,
          signer: new VoidSigner('MY ADDRESS').connect({
            getNetwork: () => Promise.resolve({ chainId: 11155111n }),
          } as any),
        },
        { nonce: 1 },
      ),
    ).rejects.toThrow(
      'invalid asset and chain combination: {"chain":"Ethereum","asset":"BTC"}',
    );
  });

  it('rejects if destination asset and chain are not valid', async () => {
    await expect(
      executeSwap(
        {
          amount: '1',
          destAsset: Assets.DOT,
          destChain: Chains.Bitcoin,
          destAddress: TESTNET_BTC_ADDRESS,
          srcAsset: Assets.ETH,
          srcChain: Chains.Ethereum,
        },
        {
          network: ChainflipNetworks.sisyphos,
          signer: new VoidSigner('MY ADDRESS').connect({
            getNetwork: () => Promise.resolve({ chainId: 11155111n }),
          } as any),
        },
        { nonce: 1 },
      ),
    ).rejects.toThrow(
      'invalid asset and chain combination: {"chain":"Bitcoin","asset":"DOT"}',
    );
  });

  it('rejects an invalid destination address', async () => {
    await expect(
      executeSwap(
        {
          amount: '1',
          destAsset: Assets.BTC,
          destChain: Chains.Bitcoin,
          destAddress: 'invalid-btc-address',
          srcAsset: Assets.ETH,
          srcChain: Chains.Ethereum,
        },
        {
          network: ChainflipNetworks.sisyphos,
          signer: new VoidSigner('MY ADDRESS').connect({
            getNetwork: () => Promise.resolve({ chainId: 11155111n }),
          } as any),
        },
        { nonce: 1 },
      ),
    ).rejects.toThrow(
      'Address "invalid-btc-address" is not a valid "Bitcoin" address',
    );
  });

  it('rejects if the source chain is not an evm chain', async () => {
    await expect(
      executeSwap(
        {
          amount: '1',
          destAsset: Assets.BTC,
          destChain: Chains.Bitcoin,
          destAddress: TESTNET_BTC_ADDRESS,
          srcAsset: Assets.DOT,
          srcChain: Chains.Polkadot,
        },
        {
          network: ChainflipNetworks.sisyphos,
          signer: new VoidSigner('MY ADDRESS').connect({
            getNetwork: () => Promise.resolve({ chainId: 11155111n }),
          } as any),
        },
        { nonce: 1 },
      ),
    ).rejects.toThrow('Chain Polkadot is not an evm chain');
  });

  it('rejects if the signer is connected to the wrong evm chain', async () => {
    await expect(
      executeSwap(
        {
          amount: '1',
          destAsset: Assets.BTC,
          destChain: Chains.Bitcoin,
          destAddress: TESTNET_BTC_ADDRESS,
          srcAsset: Assets.ETH,
          srcChain: Chains.Ethereum,
        },
        {
          network: ChainflipNetworks.sisyphos,
          signer: new VoidSigner('MY ADDRESS').connect({
            getNetwork: () => Promise.resolve({ chainId: 404n }),
          } as any),
        },
        { nonce: 1 },
      ),
    ).rejects.toThrow(
      'Signer is connected to unexpected evm chain (expected: 11155111, got: 404)',
    );
  });

  it('rejects if the destination chain is not an evm chain for a call', async () => {
    await expect(
      executeSwap(
        {
          amount: '1',
          destAsset: Assets.BTC,
          destChain: Chains.Bitcoin,
          destAddress: TESTNET_BTC_ADDRESS,
          srcAsset: Assets.ETH,
          srcChain: Chains.Ethereum,
          ccmMetadata: { message: '0xdeadc0de', gasBudget: '101' },
        },
        {
          network: ChainflipNetworks.sisyphos,
          signer: new VoidSigner('MY ADDRESS').connect({
            getNetwork: () => Promise.resolve({ chainId: 11155111n }),
          } as any),
        },
        { nonce: 1 },
      ),
    ).rejects.toThrow('Chain Bitcoin is not an evm chain');
  });

  it.each([
    {
      srcAsset: Assets.ETH,
      srcChain: Chains.Ethereum,
      destAsset: Assets.FLIP,
      destChain: Chains.Ethereum,
      destAddress: ETH_ADDRESS,
      ccmMetadata: { message: '0xdeadc0de', gasBudget: '101' },
    },
    {
      srcAsset: Assets.ETH,
      srcChain: Chains.Ethereum,
      destAsset: Assets.USDC,
      destChain: Chains.Ethereum,
      destAddress: ETH_ADDRESS,
      ccmMetadata: { message: '0xdeadc0de', gasBudget: '101' },
    },
  ])('submits a native call (%p)', async (params) => {
    const wait = jest
      .fn()
      .mockResolvedValue({ status: 1, hash: 'hello world' });
    const callSpy = jest
      .spyOn(MockVault.prototype, 'xCallNative')
      .mockResolvedValue({ wait });

    expect(
      await executeSwap(
        {
          amount: '1',
          ...params,
        } as ExecuteSwapParams,
        {
          network: ChainflipNetworks.sisyphos,
          signer: new VoidSigner('MY ADDRESS').connect({
            getNetwork: () => Promise.resolve({ chainId: 11155111n }),
          } as any),
        },
        {},
      ),
    ).toStrictEqual({ status: 1, hash: 'hello world' });
    expect(wait).toHaveBeenCalledWith(undefined);
    expect(callSpy.mock.calls).toMatchSnapshot();
  });

  it.each([
    ...(
      [
        {
          srcAsset: Assets.FLIP,
          srcChain: Chains.Ethereum,
          ccmMetadata: { message: '0xdeadc0de', gasBudget: '101' },
        },
        {
          srcAsset: Assets.USDC,
          srcChain: Chains.Ethereum,
          ccmMetadata: { message: '0xdeadc0de', gasBudget: '101' },
        },
      ] as const
    ).flatMap(
      (src) =>
        [
          {
            destAsset: Assets.ETH,
            destChain: Chains.Ethereum,
            destAddress: ETH_ADDRESS,
            ...src,
          },
        ] as const,
    ),
  ])('submits a token call (%p)', async (params) => {
    const wait = jest
      .fn()
      .mockResolvedValue({ status: 1, hash: 'hello world' });
    const approveSpy = jest
      .spyOn(MockERC20.prototype, 'approve')
      .mockResolvedValue({ wait });
    const callSpy = jest
      .spyOn(MockVault.prototype, 'xCallToken')
      .mockResolvedValue({ wait });
    const allowanceSpy = jest
      .spyOn(MockERC20.prototype, 'allowance')
      .mockResolvedValueOnce(BigInt(Number.MAX_SAFE_INTEGER - 1));

    expect(
      await executeSwap(
        {
          amount: '1',
          ...params,
        },
        {
          network: 'sisyphos',
          signer: new VoidSigner('MY ADDRESS').connect({
            getNetwork: () => Promise.resolve({ chainId: 11155111n }),
          } as any),
        },
        {},
      ),
    ).toStrictEqual({ status: 1, hash: 'hello world' });
    expect(wait).toHaveBeenCalledWith(undefined);
    expect(callSpy).toHaveBeenCalled();
    expect(callSpy.mock.calls).toMatchSnapshot();
    expect(allowanceSpy.mock.calls).toMatchSnapshot();
    expect(approveSpy).not.toHaveBeenCalled();
  });
});
