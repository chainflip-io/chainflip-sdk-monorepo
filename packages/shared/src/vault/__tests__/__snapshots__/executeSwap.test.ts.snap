// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`executeSwap accepts a nonce ('1') 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    [],
    {
      "nonce": "1",
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap accepts a nonce (1) 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    [],
    {
      "nonce": 1,
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap accepts a nonce (1n) 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    [],
    {
      "nonce": 1n,
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap can be invoked with localnet options 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    "0x456",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap can be invoked with localnet options 2`] = `
[
  [
    "MY ADDRESS",
    "0x123",
  ],
]
`;

exports[`executeSwap submits a native swap ({"destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2", "destAsset": "FLIP", "destChain": "Ethereum"}) 1`] = `
[
  [
    1,
    "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2",
    2,
    [],
    {
      "nonce": undefined,
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap submits a native swap ({"destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2", "destAsset": "USDC", "destChain": "Ethereum"}) 1`] = `
[
  [
    1,
    "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2",
    3,
    [],
    {
      "nonce": undefined,
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap submits a native swap ({"destAddress": "5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX", "destAsset": "DOT", "destChain": "Polkadot"}) 1`] = `
[
  [
    2,
    "0x841226ea070c9577979ca2e854130fbe3253853c13c05943e09908312950275d",
    4,
    [],
    {
      "nonce": undefined,
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap submits a native swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin"}) 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    [],
    {
      "nonce": undefined,
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap submits a native swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin"}) 2`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    [],
    {
      "nonce": undefined,
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2", "destAsset": "ETH", "destChain": "Ethereum", "srcAsset": "FLIP"}) 1`] = `
[
  [
    1,
    "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2",
    1,
    "0x8e3D3250470314f2e8C1eC6062Df43B5d67B5b7d",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2", "destAsset": "ETH", "destChain": "Ethereum", "srcAsset": "FLIP"}) 2`] = `
[
  [
    "MY ADDRESS",
    "0xBFF0d5051bC984ee6fF052b26ADBaFB5F1cDF9d4",
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2", "destAsset": "ETH", "destChain": "Ethereum", "srcAsset": "USDC"}) 1`] = `
[
  [
    1,
    "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2",
    1,
    "0x07865c6E87B9F70255377e024ace6630C1Eaa37F",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2", "destAsset": "ETH", "destChain": "Ethereum", "srcAsset": "USDC"}) 2`] = `
[
  [
    "MY ADDRESS",
    "0xBFF0d5051bC984ee6fF052b26ADBaFB5F1cDF9d4",
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX", "destAsset": "DOT", "destChain": "Polkadot", "srcAsset": "FLIP"}) 1`] = `
[
  [
    2,
    "0x841226ea070c9577979ca2e854130fbe3253853c13c05943e09908312950275d",
    4,
    "0x8e3D3250470314f2e8C1eC6062Df43B5d67B5b7d",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX", "destAsset": "DOT", "destChain": "Polkadot", "srcAsset": "FLIP"}) 2`] = `
[
  [
    "MY ADDRESS",
    "0xBFF0d5051bC984ee6fF052b26ADBaFB5F1cDF9d4",
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX", "destAsset": "DOT", "destChain": "Polkadot", "srcAsset": "USDC"}) 1`] = `
[
  [
    2,
    "0x841226ea070c9577979ca2e854130fbe3253853c13c05943e09908312950275d",
    4,
    "0x07865c6E87B9F70255377e024ace6630C1Eaa37F",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX", "destAsset": "DOT", "destChain": "Polkadot", "srcAsset": "USDC"}) 2`] = `
[
  [
    "MY ADDRESS",
    "0xBFF0d5051bC984ee6fF052b26ADBaFB5F1cDF9d4",
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin", "srcAsset": "FLIP"}) 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    "0x8e3D3250470314f2e8C1eC6062Df43B5d67B5b7d",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin", "srcAsset": "FLIP"}) 2`] = `
[
  [
    "MY ADDRESS",
    "0xBFF0d5051bC984ee6fF052b26ADBaFB5F1cDF9d4",
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin", "srcAsset": "FLIP"}) 3`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    "0x8e3D3250470314f2e8C1eC6062Df43B5d67B5b7d",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin", "srcAsset": "FLIP"}) 4`] = `
[
  [
    "MY ADDRESS",
    "0xBFF0d5051bC984ee6fF052b26ADBaFB5F1cDF9d4",
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin", "srcAsset": "USDC"}) 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    "0x07865c6E87B9F70255377e024ace6630C1Eaa37F",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin", "srcAsset": "USDC"}) 2`] = `
[
  [
    "MY ADDRESS",
    "0xBFF0d5051bC984ee6fF052b26ADBaFB5F1cDF9d4",
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin", "srcAsset": "USDC"}) 3`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    "0x07865c6E87B9F70255377e024ace6630C1Eaa37F",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin", "srcAsset": "USDC"}) 4`] = `
[
  [
    "MY ADDRESS",
    "0xBFF0d5051bC984ee6fF052b26ADBaFB5F1cDF9d4",
  ],
]
`;

exports[`executeSwap submits a token swap with sufficient approval 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    "0x8e3D3250470314f2e8C1eC6062Df43B5d67B5b7d",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap submits a token swap with sufficient approval 2`] = `
[
  [
    "MY ADDRESS",
    "0xBFF0d5051bC984ee6fF052b26ADBaFB5F1cDF9d4",
  ],
]
`;
